core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -171,31 +317,77 @@ private void applySingleEntry(JournalEntry entry) {
+  private File createTempSnapshotFile() throws IOException {
+    File tempDir = new File(mStorage.getSmDir().getParentFile(), "tmp");
+    if (!tempDir.isDirectory() && !tempDir.mkdir()) {
+      throw new IOException(
+          "Cannot create temporary snapshot directory at " + tempDir.getAbsolutePath());
+    }
+    return File.createTempFile("raft_snapshot_" + System.currentTimeMillis() + "_",
+        ".dat", tempDir);
+  }
+
+  /**
+   * Takes a snapshot of local state machine.
+   * @return the index of last included entry, or {@link RaftLog#INVALID_LOG_INDEX} if it fails
+   */
+  public long takeLocalSnapshot() {
     // Snapshot format is [snapshotId, name1, bytes1, name2, bytes2, ...].
     if (mClosed) {
-      return;
+      return RaftLog.INVALID_LOG_INDEX;
     }
     LOG.debug("Calling snapshot");
     Preconditions.checkState(!mSnapshotting, "Cannot call snapshot multiple times concurrently");
     mSnapshotting = true;
     mLastSnapshotStartTime = System.currentTimeMillis();
     long snapshotId = mNextSequenceNumberToRead - 1;
-    try (SnapshotWriterStream sws = new SnapshotWriterStream(writer)) {
-      writer.writeLong(snapshotId);
-      JournalUtils.writeToCheckpoint(sws, getStateMachines());
+    TermIndex last = getLastAppliedTermIndex();
+    File tempFile;
+    try {
+      tempFile = createTempSnapshotFile();
+    } catch (IOException e) {
+      LogUtils.warnWithException(LOG, "Failed to create temp snapshot file", e);
+      return RaftLog.INVALID_LOG_INDEX;
+    }
+    LOG.info("Taking a snapshot to file {}", tempFile);
+    final File snapshotFile = mStorage.getSnapshotFile(last.getTerm(), last.getIndex());
+    try (DataOutputStream outputStream = new DataOutputStream(new FileOutputStream(tempFile))) {
+      outputStream.writeLong(snapshotId);
+      JournalUtils.writeToCheckpoint(outputStream, getStateMachines());
+    } catch (Exception e) {
+      tempFile.delete();
+      LogUtils.warnWithException(LOG, "Failed to take snapshot: {}", snapshotId, e);
+      return RaftLog.INVALID_LOG_INDEX;
+    }
+    try {
+      final MD5Hash digest = MD5FileUtil.computeMd5ForFile(tempFile);
+      LOG.info("Saving digest for snapshot file {}", snapshotFile);
+      MD5FileUtil.saveMD5File(snapshotFile, digest);
+      LOG.info("Renaming a snapshot file {} to {}", tempFile, snapshotFile);
+      if (!tempFile.renameTo(snapshotFile)) {
+        tempFile.delete();
+        LOG.warn("Failed to rename snapshot from {} to {}", tempFile, snapshotFile);
+        return RaftLog.INVALID_LOG_INDEX;
+      }
+      LOG.info("Completed snapshot up to SN {} in {}ms", snapshotId,
+          System.currentTimeMillis() - mLastSnapshotStartTime);
     } catch (Exception e) {
-      ProcessUtils.fatalError(LOG, e, "Failed to take snapshot: %s", snapshotId);
-      throw new RuntimeException(e);
+      tempFile.delete();
+      LogUtils.warnWithException(LOG, "Failed to take snapshot: {}", snapshotId, e);
```

Comment:
Include filename? Also, this message is identical to the previous section. Can we distinguish them?

Reply:
Updated.